### PR TITLE
[TimeLock Corruption Detection] | Integration tests

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
Adds integration tests for corruption detection flow.

**Implementation Description (bullets)**:

- Adds integration tests in CorruptionDetectionIntegrationTest

- Pulls up common set up methods to TimeLockCorruptionTestSetup

**Testing (What was existing testing like?  What have you done to improve it?)**:
All tests

**Concerns (what feedback would you like?)**:
More test cases?

**Where should we start reviewing?**:
CorruptionDetectionIntegrationTest

**Priority (whenever / two weeks / yesterday)**:
Tomorrow 🏃‍♀️

### CODE REVIEW

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import com.palantir.paxos.NamespaceAndUseCase;
+import java.util.List;
+import java.util.stream.IntStream;
+import org.junit.Test;
+
+/**
+ * This class performs integration tests by inducing and detecting corruption in one or more series.
+ * Note - All tests only induce and detect one type of corruption check violation -> ACCEPTED_VALUE_GREATER_THAN_LEARNED
+ */
+public class CorruptionDetectionIntegrationTest extends TimeLockCorruptionTestSetup {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:34]: nit: final class?

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Test
+    public void detectCorruptionForLogAtSeqAtBatchEnd() {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:37]: nit: test name is hard to parse

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        assertDetectedViolations(ImmutableSet.of(CorruptionCheckViolation.ACCEPTED_VALUE_GREATER_THAN_LEARNED));
+    }
+
+    @Test
+    public void detectCorruptionForLogSeqInLaterBatches() {
+        writeLogsOnLocalAndRemote(1, 1000);
+        induceGreaterAcceptedValueCorruption(599);
+
+        // No signs of corruption in the first batch
+        SetMultimap<CorruptionCheckViolation, NamespaceAndUseCase> violationsToNamespaceToUseCaseMultimap =
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:51]: prefer `Multimap`

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ */
+public class CorruptionDetectionIntegrationTest extends TimeLockCorruptionTestSetup {
+
+    @Test
+    public void detectCorruptionForLogAtSeqAtBatchEnd() {
+        // We write logs in range - [1, 500]. The first range of sequences for corruption detection = [0, 499] since
+        // this range is computed from INITIAL_PROGRESS = -1.
+        writeLogsOnLocalAndRemote(1, 500);
+        induceGreaterAcceptedValueCorruption(499);
+        assertDetectedViolations(ImmutableSet.of(CorruptionCheckViolation.ACCEPTED_VALUE_GREATER_THAN_LEARNED));
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:42]: minor nit: `assertViolationsDetected`

Comment (Jolyon-S) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:42]: also, split into two methods: some violations (and maybe can be specific), or no violation.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.sql.DataSource;
+import org.immutables.value.Value;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+
+public abstract class TimeLockCorruptionTestSetup {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:52]: chatted offline - I personally don't like this structure of using abstract classes; would prefer if this was a final class that you call methods on directly.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+    protected List<StateLogComponents> createStatLogComponentsForNamespaceAndUseCase(
+            NamespaceAndUseCase namespaceAndUseCase) {
+        return ImmutableList.of(
+                createLogComponentsForServer(localDataSource, namespaceAndUseCase),
+                createLogComponentsForServer(remoteDataSource1, namespaceAndUseCase),
+                createLogComponentsForServer(remoteDataSource2, namespaceAndUseCase));
+    }
+
+    // utils
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:96]: please remove this

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    protected SetMultimap<CorruptionCheckViolation, NamespaceAndUseCase> getViolationsToNamespaceToUseCaseMultimap() {
+        List<CompletePaxosHistoryForNamespaceAndUseCase> historyForAll = paxosLogHistoryProvider.getHistory();
+        return HistoryAnalyzer.corruptionHealthReportForHistory(historyForAll).violatingStatusesToNamespaceAndUseCase();
+    }
+
+    protected void assertDetectedViolations(Set<CorruptionCheckViolation> detectedViolations) {
+        assertDetectedViolations(detectedViolations, ImmutableSet.of(NAMESPACE_AND_USE_CASE));
+    }
+
+    protected void assertDetectedViolations(
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:169]: IntelliJ is your friend; these probably? don't have to be protected

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/Constants.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package com.palantir.timelock;
+
+import com.palantir.paxos.Client;
+import com.palantir.paxos.ImmutableNamespaceAndUseCase;
+import com.palantir.paxos.NamespaceAndUseCase;
+
+public class Constants {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/Constants.java:23]: nit: can we call this `TimelockCorruptionTestConstants` or similar? I know it is in the test project, but I'm wary of such a generic name here

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+import com.palantir.paxos.NamespaceAndUseCase;
+import com.palantir.timelock.corruption.detection.TimeLockCorruptionTestSetup.StateLogComponents;
+import java.util.List;
+import java.util.stream.IntStream;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * This class performs integration tests by inducing and detecting corruption in one or more series.
+ * Note - All tests only induce and detect one type of corruption check violation -> ACCEPTED_VALUE_GREATER_THAN_LEARNED
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:31]: the javadoc is a note, of sorts. I'd recommend refactoring this line as follows:
```suggestion
 * All tests only induce and detect ACCEPTED_VALUE_GREATER_THAN_LEARNED corruption.
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+ */
+public final class CorruptionDetectionIntegrationTest {
+    @Rule
+    public TimeLockCorruptionDetectionHelper helper = new TimeLockCorruptionDetectionHelper();
+
+    @Test
+    public void detectCorruptionForLogAtBatchEnd() {
+        // We write logs in range - [1, 500]. The first range of sequences for corruption detection = [0, 499] since
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:39]: nit: no need to hyphenate here.
```suggestion
        // We write logs in range [1, 500]. The first range of sequences for corruption detection = [0, 499] since
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+                helper.createStatLogComponentsForNamespaceAndUseCase(namespaceAndUseCase), 1, 500);
+    }
+
+    private void corruptSeries(Integer namespaceAndUseCaseIndex) {
+        NamespaceAndUseCase namespaceAndUseCase = namespaceAndUseCaseForIndex(namespaceAndUseCaseIndex);
+        List<StateLogComponents> components = helper.createStatLogComponentsForNamespaceAndUseCase(namespaceAndUseCase);
+        helper.induceGreaterAcceptedValueCorruption(components.get(0), 499);
+    }
+
+    private NamespaceAndUseCase namespaceAndUseCaseForIndex(Integer ind) {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:107]: nit:
```suggestion
    private static NamespaceAndUseCase namespaceAndUseCaseForIndex(Integer ind) {
```

Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:107]: nit:
```suggestion
    private NamespaceAndUseCase namespaceAndUseCaseForIndex(int index) {
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                ImmutableSet.of(namespaceAndUseCaseForIndex(6), namespaceAndUseCaseForIndex(7)));
+    }
+
+    private void createSeriesWithPaxosLogs(Integer namespaceAndUseCaseIndex) {
+        NamespaceAndUseCase namespaceAndUseCase = namespaceAndUseCaseForIndex(namespaceAndUseCaseIndex);
+        helper.writeLogsOnLocalAndRemote(
+                helper.createStatLogComponentsForNamespaceAndUseCase(namespaceAndUseCase), 1, 500);
+    }
+
+    private void corruptSeries(Integer namespaceAndUseCaseIndex) {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:101]: nit:
```suggestion
    private void corruptSeries(int namespaceAndUseCaseIndex) {
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        // We create 7 series and write logs to each of these in the range [1, 500]. We then corrupt series 6 & 7.
+        IntStream.rangeClosed(1, 7).boxed().forEach(this::createSeriesWithPaxosLogs);
+        IntStream.rangeClosed(6, 7).boxed().forEach(this::corruptSeries);
+
+        helper.assertViolationsDetectedForNamespaceAndUseCases(
+                ImmutableSet.of(CorruptionCheckViolation.ACCEPTED_VALUE_GREATER_THAN_LEARNED),
+                ImmutableSet.of(namespaceAndUseCaseForIndex(6), namespaceAndUseCaseForIndex(7)));
+    }
+
+    private void createSeriesWithPaxosLogs(Integer namespaceAndUseCaseIndex) {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/CorruptionDetectionIntegrationTest.java:95]: nit:
```suggestion
    private void createSeriesWithPaxosLogs(int namespaceAndUseCaseIndex) {
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    void assertViolationsDetected(Set<CorruptionCheckViolation> detectedViolations) {
+        assertViolationsDetectedForNamespaceAndUseCases(
+                detectedViolations, ImmutableSet.of(Constants.DEFAULT_NAMESPACE_AND_USE_CASE));
+    }
+
+    void assertViolationsDetectedForNamespaceAndUseCases(
+            Set<CorruptionCheckViolation> detectedViolations,
+            Set<NamespaceAndUseCase> namespaceAndUseCasesWithViolation) {
+
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:97]: nit: remove empty line
```suggestion
```

Comment (Jolyon-S) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:97]: With this method, either 1 or 0 detected violations are passed, which I'm not a fan of at all. I'd prefer if you separate out to two methods: one that takes a single violation and confirms it is there, and a second where the logic is essentially:
```java
assertThat(getViolationsToNamespaceToUseCaseMultimap()).isEmpty();
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        Multimap<CorruptionCheckViolation, NamespaceAndUseCase> violationsToNamespaceToUseCaseMultimap =
+                getViolationsToNamespaceToUseCaseMultimap();
+
+        assertThat(violationsToNamespaceToUseCaseMultimap.keySet()).hasSameElementsAs(detectedViolations);
+        assertThat(violationsToNamespaceToUseCaseMultimap.values())
+                .hasSameElementsAs(namespaceAndUseCasesWithViolation);
+    }
+
+    private Multimap<CorruptionCheckViolation, NamespaceAndUseCase> getViolationsToNamespaceToUseCaseMultimap() {
+        List<CompletePaxosHistoryForNamespaceAndUseCase> historyForAll = getHistory();
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:107]: you can inline this variable - the method call is actually shorter than the variable!

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+import com.palantir.timelock.history.models.CompletePaxosHistoryForNamespaceAndUseCase;
+import com.palantir.timelock.history.utils.PaxosSerializationTestUtils;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class TimeLockCorruptionDetectionHelper implements TestRule {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:36]: IntelliJ tells me that there are a number of methods that can be private - please make them private, then re-order so that they are after the non-private ones.

Also, there are definitely some methods that can be static - for example `writeLogsOnServer`. Please make static the methods that can be.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        writeLogsOnLocalAndRemote(
+                timeLockCorruptionTestSetup.getDefaultServerList(), startingLogSeq, latestLogSequence);
+    }
+
+    Set<PaxosValue> writeLogsOnServer(StateLogComponents server, int start, int end) {
+        return PaxosSerializationTestUtils.writeToLogs(server.acceptorLog(), server.learnerLog(), start, end);
+    }
+
+    void writeLogsOnLocalAndRemote(List<StateLogComponents> servers, int startingLogSeq, int latestLogSequence) {
+        servers.stream().forEach(server -> writeLogsOnServer(server, startingLogSeq, latestLogSequence));
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:53]: We discussed - remove stream()
```suggestion
        servers.forEach(server -> writeLogsOnServer(server, startingLogSeq, latestLogSequence));
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return RuleChain.outerRule(tempFolder)
+                .around(new ExternalResource() {
+                    @Override
+                    protected void before() {
+                        try {
+                            setup();
+                        } catch (Throwable throwable) {
+                            // no op
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:66]: Interesting - if we get an exception on startup, we don't want to do anything? I'd have thought we'd do:
```java
throw new RuntimeException(throwable, "failed on startup");
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+    private PaxosLogHistoryProvider paxosLogHistoryProvider() {
+        return new PaxosLogHistoryProvider(
+                localDataSource,
+                defaultRemoteServerList.stream()
+                        .map(StateLogComponents::dataSource)
+                        .map(this::getHistoryProviderResource)
+                        .collect(Collectors.toList()));
+    }
+
+    public List<StateLogComponents> createStatLogComponentsForNamespaceAndUseCase(
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:101]: Same as elsewhere: this and other methods don't need to be public.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+
+    public List<StateLogComponents> getDefaultRemoteServerList() {
+        return defaultRemoteServerList;
+    }
+
+    private StateLogComponents createLogComponentsForServer(DataSource dataSource) {
+        return createLogComponentsForServer(dataSource, Constants.DEFAULT_NAMESPACE_AND_USE_CASE);
+    }
+
+    private StateLogComponents createLogComponentsForServer(
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:132]: nit: `static`?

Comment (Jolyon-S) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:132]: there are others in this class too

**Review by Jolyon-S — COMMENTED** [review]
There are a fair number of minor style things here, but as far as I can tell, the semantics are solid enough.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java (hunk 19 above)
Comment (jeremyk-91) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionTestSetup.java:101]: also: `StatLog`?

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.Set;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class TimeLockCorruptionDetectionHelper implements TestRule {
+    private TimeLockCorruptionTestSetup timeLockCorruptionTestSetup = new TimeLockCorruptionTestSetup();
+
+    Set<PaxosValue> writeLogsOnDefaultLocalServer(int start, int end) {
+        return writeLogsOnServer(timeLockCorruptionTestSetup.getDefaultLocalServer(), start, end);
```
Comment (sudiksha27) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:40]: name - add 'inclusive'

**Review by jeremyk-91 — COMMENTED** [review]
Logical test coverage is good. I'm in sync with @Jolyon-S's comments.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Set;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class TimeLockCorruptionDetectionHelper implements TestRule {
+    private TimeLockCorruptionTestSetup timeLockCorruptionTestSetup = new TimeLockCorruptionTestSetup();
+
+    void writeLogsOnDefaultLocalServer(int startInclusive, int endInclusive) {
+        writeLogsOnServer(timeLockCorruptionTestSetup.getDefaultLocalServer(), startInclusive, endInclusive);
```
Comment (sudiksha27) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:39]: It is only used in one place but I think this signature makes sense.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java (hunk 16 above)
Comment (sudiksha27) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:35]: Made the methods package-private. 

Done.

**Review by sudiksha27 — COMMENTED** [review]
All comments addressed. Would be good to have a 👍

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/HistoryAnalyzerTest.java [hunk 23]
```diff
@@ -18,178 +18,72 @@
...
-                        .violatingStatusesToNamespaceAndUseCase();
-        assertThat(setMultimap.keySet()).containsExactly(CorruptionCheckViolation.VALUE_LEARNED_WITHOUT_QUORUM);
+        helper.assertViolationDetected(CorruptionCheckViolation.VALUE_LEARNED_WITHOUT_QUORUM);
     }
 
     @Test
     public void detectCorruptionIfLearnedValueIsNotTheGreatestAcceptedValue() {
-        writeLogsOnServer(localStateLogComponents, 1, 5);
-        remoteStateLogComponents.forEach(server -> writeLogsOnServer(server, 1, 5));
+        helper.writeLogsOnDefaultLocalAndRemote(1, 10);
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/HistoryAnalyzerTest.java:76]: As far as I can tell, every test uses the params `1, 10`. Perhaps you should mix this up a bit? Can these numbers be large (e.g. 10k)? Might help to catch any funny errors to have at least one test with large numbers - if that makes sense in prod.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+import com.palantir.timelock.history.models.CompletePaxosHistoryForNamespaceAndUseCase;
+import com.palantir.timelock.history.utils.PaxosSerializationTestUtils;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class TimeLockCorruptionDetectionHelper implements TestRule {
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:35]: nit: reorder the methods so that public is at the top, private at the bottom (and if it makes sense to group static methods, feel free to do that too).

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/history/utils/HistoryQueries.java [hunk 25]
```diff
@@ -16,20 +16,16 @@
...
 public class HistoryQueries {
-    public static final Client DEFAULT_CLIENT = Client.of("client");
-    public static final String DEFAULT_USE_CASE = "useCase";
-    public static final NamespaceAndUseCase DEFAULT_NAMESPACE_AND_USE_CASE =
-            ImmutableNamespaceAndUseCase.of(DEFAULT_CLIENT, DEFAULT_USE_CASE);
 
     public static HistoryQuery unboundedHistoryQuerySinceSeq(long seqLowerBound) {
-        return unboundedHistoryQuerySinceSeqForNamespaceAndUseCase(DEFAULT_NAMESPACE_AND_USE_CASE, seqLowerBound);
+        return unboundedHistoryQuerySinceSeqForNamespaceAndUseCase(
+                TimelockCorruptionTestConstants.DEFAULT_NAMESPACE_AND_USE_CASE, seqLowerBound);
```
Comment (Jolyon-S) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/history/utils/HistoryQueries.java:28]: nit: In other tests, you statically imported this. Either do the same here, or remove the static imports elsewhere.

**Review by Jolyon-S — COMMENTED** [review]
A couple of nits, but otherwise looks good to me! @jeremyk-91 do you want to do a pass before it merges? I don't mind either way.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/HistoryAnalyzerTest.java (hunk 23 above)
Comment (sudiksha27) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/HistoryAnalyzerTest.java]: First part is done. For the second part I can add a (stress?) test in CorruptionDetectionIntegrationTest.

**Review by Jolyon-S — APPROVED** [review]
I trust that you'll tidy up any nits from my last review, otherwise looks good.

**Review by jeremyk-91 — APPROVED** [review]
(I reviewed the logic of the tests, and trust @Jolyon-S on the style/factoring of this.)
